File: app/src/main/java/com/dumuzeyn/mp3player/MainScreenView.java
package com.dumuzeyn.mp3player;

import android.content.Context;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import com.dumuzeyn.mp3player.ui.layout.ResponsiveLayoutController;

/** Builds and owns the stable top-level view tree of the main screen. */
final class MainScreenView {
    interface Callbacks {
        void buildHeader(LinearLayout page);

        void buildTabs(LinearLayout page);

        void buildMiniPlayer(FrameLayout root);

        void onContentScrolled();

        ParticleEffectsView createParticles();

        PlayerGradientBackground.Config gradientConfig();
    }

    static final class Appearance {
        final int solidColor;
        final int backgroundMode;
        final int gradientStart;
        final int gradientEnd;
        final String mediaUri;
        final int mediaBlur;

        Appearance(int solidColor, int backgroundMode, int gradientStart, int gradientEnd,
                String mediaUri, int mediaBlur) {
            this.solidColor = solidColor;
            this.backgroundMode = backgroundMode;
            this.gradientStart = gradientStart;
            this.gradientEnd = gradientEnd;
            this.mediaUri = mediaUri == null ? "" : mediaUri;
            this.mediaBlur = mediaBlur;
        }
    }

    static final class References {
        final FrameLayout root;
        final LinearLayout page;
        final FrameLayout contentHost;
        final ScrollView contentScroll;
        final LinearLayout contentList;
        final FrameLayout overlayHost;
        final ParticleEffectsView particles;

        References(FrameLayout root, LinearLayout page, FrameLayout contentHost,
                ScrollView contentScroll, LinearLayout contentList,
                FrameLayout overlayHost, ParticleEffectsView particles) {
            this.root = root;
            this.page = page;
            this.contentHost = contentHost;
            this.contentScroll = contentScroll;
            this.contentList = contentList;
            this.overlayHost = overlayHost;
            this.particles = particles;
        }
    }

    private final Context context;
    private final ResponsiveLayoutController layout;
    private References references;

    MainScreenView(Context context, ResponsiveLayoutController layout) {
        this.context = context;
        this.layout = layout;
    }

    References build(Appearance appearance, Callbacks callbacks) {
        FrameLayout root = new FrameLayout(context);
        root.setBackgroundColor(appearance.solidColor);
        addBackground(root, appearance, callbacks);

        LinearLayout page = new LinearLayout(context);
        page.setOrientation(LinearLayout.VERTICAL);
        int horizontalPadding = layout.pageHorizontalPadding();
        page.setPadding(horizontalPadding, layout.pageTopPadding(), horizontalPadding, 0);
        root.addView(page, layout.mainPageParams());
        callbacks.buildHeader(page);
        callbacks.buildTabs(page);

        FrameLayout contentHost = new FrameLayout(context);
        ScrollView contentScroll = new ScrollView(context);
        LinearLayout contentList = new LinearLayout(context);
        contentList.setOrientation(LinearLayout.VERTICAL);
        contentScroll.addView(contentList, new FrameLayout.LayoutParams(-1, -2));
        contentScroll.setOnScrollChangeListener(
                (view, scrollX, scrollY, oldScrollX, oldScrollY) ->
                        callbacks.onContentScrolled());
        contentHost.addView(contentScroll, new FrameLayout.LayoutParams(-1, -1));
        page.addView(contentHost, new LinearLayout.LayoutParams(-1, 0, 1.0f));

        FrameLayout overlayHost = new FrameLayout(context);
        root.addView(overlayHost, new FrameLayout.LayoutParams(-1, -1));
        callbacks.buildMiniPlayer(root);

        ParticleEffectsView particles = callbacks.createParticles();
        root.addView(particles, new FrameLayout.LayoutParams(-1, -1));
        references = new References(root, page, contentHost, contentScroll, contentList,
                overlayHost, particles);
        return references;
    }

    FrameLayout contentHost() {
        return references == null ? null : references.contentHost;
    }

    ScrollView contentScroll() {
        return references == null ? null : references.contentScroll;
    }

    void replaceContent(ScrollView scrollView, LinearLayout content) {
        if (references == null) {
            return;
        }
        references = new References(references.root, references.page,
                references.contentHost, scrollView, content,
                references.overlayHost, references.particles);
    }

    private void addBackground(
            FrameLayout root, Appearance appearance, Callbacks callbacks) {
        if (appearance.backgroundMode == BackgroundSettingsController.MODE_GRADIENT) {
            root.addView(new PlayerGradientBackground(
                    context, callbacks.gradientConfig(),
                    appearance.gradientStart, appearance.gradientEnd),
                    new FrameLayout.LayoutParams(-1, -1));
        } else if (appearance.backgroundMode == BackgroundSettingsController.MODE_MEDIA
                && !appearance.mediaUri.isEmpty()) {
            root.addView(new BackgroundMediaView(
                    context, appearance.mediaUri, appearance.mediaBlur,
                    appearance.solidColor),
                    new FrameLayout.LayoutParams(-1, -1));
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/MainScreenView.kt
package com.dumuzeyn.mp3player

import android.content.Context
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.ScrollView
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.dumuzeyn.mp3player.ui.layout.ResponsiveLayoutController

/** Builds and owns the stable top-level view tree of the main screen. */
internal class MainScreenView(
    private val context: Context,
    private val layout: ResponsiveLayoutController,
) {
    interface Callbacks {
        fun buildTabs(page: LinearLayout)
        fun buildMiniPlayer(root: FrameLayout)
        fun onContentScrolled()
        fun createParticles(): ParticleEffectsView
        fun gradientConfig(): PlayerGradientBackground.Config
    }

    class Appearance(
        val solidColor: Int,
        val backgroundMode: Int,
        val gradientStart: Int,
        val gradientEnd: Int,
        mediaUri: String?,
        val mediaBlur: Int,
    ) {
        val mediaUri = mediaUri.orEmpty()
    }

    class References(
        val root: FrameLayout,
        val page: LinearLayout,
        val contentHost: FrameLayout,
        val contentScroll: ScrollView,
        val contentList: LinearLayout,
        val overlayHost: FrameLayout,
        val particles: ParticleEffectsView,
    )

    private var references: References? = null

    fun build(appearance: Appearance, callbacks: Callbacks): References {
        val root = FrameLayout(context).apply {
            setBackgroundColor(appearance.solidColor)
            ViewCompat.setOnApplyWindowInsetsListener(this) { view, insets ->
                val bars = insets.getInsets(WindowInsetsCompat.Type.systemBars() or
                    WindowInsetsCompat.Type.displayCutout())
                view.setPadding(bars.left, bars.top, bars.right, bars.bottom)
                insets
            }
        }
        addBackground(root, appearance, callbacks)

        val page = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            clipChildren = false
            clipToPadding = false
            val horizontalPadding = layout.pageHorizontalPadding()
            setPadding(horizontalPadding, layout.pageTopPadding(), horizontalPadding, 0)
        }
        root.addView(page, layout.mainPageParams())
        callbacks.buildTabs(page)

        val contentHost = FrameLayout(context).apply {
            clipChildren = false
            clipToPadding = false
        }
        val contentScroll = ScrollView(context)
        val contentList = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            val clearance = layout.contentScrollbarClearance()
            setPadding(clearance, 0, clearance, 0)
        }
        contentScroll.addView(contentList, FrameLayout.LayoutParams(-1, -2))
        contentScroll.setOnScrollChangeListener { _, _, _, _, _ -> callbacks.onContentScrolled() }
        contentHost.addView(contentScroll, FrameLayout.LayoutParams(-1, -1))
        page.addView(contentHost, LinearLayout.LayoutParams(-1, 0, 1f))

        val overlayHost = FrameLayout(context)
        root.addView(overlayHost, FrameLayout.LayoutParams(-1, -1))
        callbacks.buildMiniPlayer(root)

        val particles = callbacks.createParticles()
        root.addView(particles, FrameLayout.LayoutParams(-1, -1))
        return References(
            root,
            page,
            contentHost,
            contentScroll,
            contentList,
            overlayHost,
            particles,
        ).also { references = it }
    }

    fun contentHost(): FrameLayout? = references?.contentHost

    fun contentScroll(): ScrollView? = references?.contentScroll

    fun replaceContent(scrollView: ScrollView, content: LinearLayout) {
        val current = references ?: return
        references = References(
            current.root,
            current.page,
            current.contentHost,
            scrollView,
            content,
            current.overlayHost,
            current.particles,
        )
    }

    private fun addBackground(root: FrameLayout, appearance: Appearance, callbacks: Callbacks) {
        when {
            appearance.backgroundMode == BackgroundSettingsController.MODE_GRADIENT -> {
                root.addView(
                    PlayerGradientBackground(
                        context,
                        callbacks.gradientConfig(),
                        appearance.gradientStart,
                        appearance.gradientEnd,
                    ),
                    FrameLayout.LayoutParams(-1, -1),
                )
            }

            appearance.backgroundMode == BackgroundSettingsController.MODE_MEDIA &&
                appearance.mediaUri.isNotEmpty() -> {
                root.addView(
                    BackgroundMediaView(
                        context,
                        appearance.mediaUri,
                        appearance.mediaBlur,
                        appearance.solidColor,
                    ),
                    FrameLayout.LayoutParams(-1, -1),
                )
            }
        }
    }
}
